### ISSUE TITLE
BeforeSuite may not be behaving as expected

### ISSUE THREAD

**Issue by amkam** [concordion/concordion#195]
Hi everyone,

We have an extension which transforms the input html fixture using some data coming from the java fixture. To prepare the given data on the java fixture object and on concordion 1.5.x we just annotated some init method with @Before and it worked.

I'm now trying to migrate to concordion 2.0 but the @Before annotation now behaves differently (in order to support examples), which is that the document is being parsed (by launching our custom extension) just before the annotated method is called and then the extension fails because it isn't fed with prepared data.

So, to fix that, I changed to @BeforeSuite but got the same result. Is this the expected behaviour or I'm doing something wrong?.

TL;DR An extension implementing DocumentParsingListener.beforeParsing method is being called before a method annotated with @BeforeSuite

**Comment by amkam** [comment]
After a deeper inspection it seems the beforeParsing method is triggered at a very early phase, nothing I put in the class even in a static block gets called before that method.

**Comment by nigelcharman** [comment]
Would you supply some code to show how the data gets from the fixture to the `DocumentParsingListener.beforeParsing` method please?

I tried a simple test and I am seeing the static initialiser called before the `beforeParsing` method on an extension.

**Comment by agseijas** [comment]
I mean: using @BeforeClass or @ClassRules or even @Rules or @BeforeSuite which as implied, they should be executed before any parsing occurs, if I'm not wrong.

In any case, lend me some time to put the code.

**Comment by amkam** [comment]
I mean: using @BeforeClass or @ClassRules or even @Rules or @BeforeSuite which as implied, they should be executed before any parsing occurs, if I'm not wrong.

In any case, lend me some time to put the code.

**Comment by amkam** [comment]
Extension class

`
import org.concordion.api.extension.ConcordionExtender;
import org.concordion.api.extension.ConcordionExtension;
import org.concordion.api.listener.DocumentParsingListener;

import nu.xom.Document;

public class SomeExtension implements ConcordionExtension, DocumentParsingListener {

	public void beforeParsing(Document document) {
		System.out.println("beforeParsing called");
	}

	public void addTo(ConcordionExtender concordionExtender) {
        concordionExtender.withDocumentParsingListener(this);
	}

}
`

Spec class
`
import org.concordion.api.extension.Extensions;
import org.concordion.integration.junit4.ConcordionRunner;
import org.junit.BeforeClass;
import org.junit.runner.RunWith;

@RunWith(ConcordionRunner.class)
@Extensions({ SomeExtension.class })
public class Spec {

	@BeforeClass
	public static void init(){
		System.out.println("init called");
	}
	
}
`

The html

`
<html xmlns:c="http://www.concordion.org/2007/concordion" xmlns:ext="urn:concordion-extensions:2010" >
   <head>
   </head>
   <body>
	<c:echo>someecho</c:echo>   
	<div></div>
   </body>
</html>
`

Dependecy concordion 2.0.0 

Once executed it shows this on console:

`
beforeParsing called
init called
`

But should be

`
init called
beforeParsing called
`

**Comment by nigelcharman** [comment]
Thanks for reporting this. As you understand, we've needed to move the specification parsing code in 2.0 so that we can determine which examples are in the specification.

While we don't make any guarantees about the order in which the specification is parsed with respect to the fixture hooks being invoked, we're seeing what we can do in this instance.

I still don't understand how your beforeParsing() method is getting access to the Fixture.

A static initialiser may provide you with a solution until we look at any changes to the method hook invocation:

```
import org.concordion.api.extension.Extensions;
import org.concordion.integration.junit4.ConcordionRunner;
import org.junit.BeforeClass;
import org.junit.runner.RunWith;

@RunWith(ConcordionRunner.class)
@Extensions({ SomeExtension.class })
public class Spec {
	
	{
		System.out.println("static init called");
	}
	
	@BeforeClass
	public static void init(){
	    System.out.println("init called");
	}
}
```

outputs:

```
static init called
beforeParsing called
init called
```

**Comment by amkam** [comment]
Well we did something like that to get it working, although some extensive refactoring had to be made. Now it just doesn't feel right to be avoiding annotations; but we can totally live with it :)

About the beforeParsing, it's relying on other extension to prepare the data I can't remember exactly the code (I don't currently have access to the code) but it's related to this

https://gist.github.com/lantoli/6b8e3ec653b2886d176c

**Comment by nigelcharman** [comment]
We're looking to move the `@BeforeSuite` annotation handling to occur before the specification is parsed. Just bouncing the idea off a few people before implementing.

**Comment by nigelcharman** [comment]
The change to run methods annotated with `@BeforeSuite` before parsing the specification has been released in Concordion v2.0.1. Please re-open this issue if this doesn't resolve it for you. thanks
